src/replication.c

```diff
@@ -2173,6 +2235,211 @@ istgt_lu_resize_volume(spec_t *spec, uint64_t size) {
 	return true;
 }
 
+#define IS_REPLICA_ID_EXIST_IN_LIST(_replica_id, _list, _no_of_items, _is_exist) 	\
+	do {										\
+		int i;									\
+		for (i=0; i < _no_of_items; i++) {					\
+			if (strncmp(_replica_id, _list[i], REPLICA_ID_LEN) == 0) {	\
+					_is_exist = true;				\
+					break;						\
+			}								\
+		}									\
+	}while(0)
+
+
+#ifdef REPLICATION
+/* is_valid_known_replica_list this function should be called by taking
+ * spec->rq_mtx lock
+ */
+static bool
+is_valid_known_replica_list(spec_t *spec, int listcnt, char **known_replica_id_list) {
+	trusty_replica_t *trusty_replica = NULL;
+	bool found_replica;
+	int i;
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+	for(i=0; i<listcnt; i++) {
+		found_replica = false;
+		TAILQ_FOREACH(trusty_replica, &spec->lu->trusty_replicas, next) {
+			if (strncmp(trusty_replica->replica_id,
+			    known_replica_id_list[i], REPLICA_ID_LEN) == 0) {
+				found_replica = true;
+				break;
+			}
+		}
+		// If known replicaid doesn't exist in trusty replica then
+		// replicaid list has wrong information
+		if (!found_replica) {
+			ISTGT_ERRLOG("replicaid (%s) doesn't exist in "
+			    "trusty replica list\n", known_replica_id_list[i]);
+			return false;
+		}
+	}
+	return true;
+}
+#endif
+
+static bool
+can_replica_remove(spec_t *spec, replica_t *removing_replica) {
+	replica_t *replica;
+
+	if (spec->healthy_rcount < spec->consistency_factor)
+		return false;
+
+	// TODO: Return true when all the replicas_id should be healthy
+	// else return false
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if (replica != removing_replica &&
+		    replica->state != ZVOL_STATUS_HEALTHY)
+			return false;
```

Comment:
Can't we remove the replica if `healthy_rcount > (spec->cf + 1)` instead of checking all the replica's state?

Reply:
No, I think if we go this approach then we will end up in having different conditions based on `replication factor` and one more gotcha with this is we can't perform replica scale down if `spec->rf is 2`.

Reply:
As discussed I will update to `healthy_rcount >= (spec->cf + 1)`

Reply:
ok. That make sense. Let's mention this condition, for replica scale-down, in PR description.

Reply:
Updated as prerequisites step to perform scaledown step